middleware/domain/source/discovery/state.cpp

```diff
@@ -139,6 +139,42 @@ namespace casual
          return false;
       }
 
-      
+      void State::Prospects::insert( const Content& content, common::strong::correlation::id correlation)
+      {
+         for( auto service : content.services)
+         {
+            services[ service].emplace_back( correlation);
+         }
+         for( auto queue : content.queues)
+         {
+            queues[ queue].emplace_back( correlation);
+         }
+      }
+
+      void State::Prospects::remove( common::strong::correlation::id correlation)
+      {
+         for( auto& service : services)
+         {
+            algorithm::container::erase( service.second, correlation);
+         }
+         algorithm::container::erase_if( services, []( auto pair) { return pair.second.empty();});
```

Comment:
You could erase the correlation in the same algorithm, hence only do one pass.

```suggestion
         algorithm::container::erase_if( services, [ &correlation]( auto& pair) 
         {
             return algorithm::container::erase( pair.second, correlation).empty();
         );
```
And you need to take the pair by reference.

Reply:
@lazan So it is safe to remove elements from _services_ when inside a range for-loop over _services_?

Reply:
Now I got it.